File: Difficulty: Easy/Strings Rotations of Each Other/strings-rotations-of-each-other.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    // Function to check if two strings are rotations of each other or not.
    bool areRotations(string &s1, string &s2) {
        // Your code here
        if (s1.size() != s2.size()) return false;
        map<int,int> mp;
        for(auto x:s1){
            mp[x]++;
        }
        for(auto x:s2){
            mp[x]--;
        }
        for(auto x:mp){
            if(x.second!=0){
                return false;
            }
        }
        string temp = s2 + s2;
        int n = s1.size();
        
        for (int i = 0; i < temp.size() - n + 1; i++) {
            int j = 0;
            while (j < n && temp[i + j] == s1[j]) {
                j++;
            }
            if (j == n) return true;
        }
        return false;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string s1;
        string s2;
        cin >> s1 >> s2;
        Solution sol;
        int result = sol.areRotations(s1, s2);
        if (result) {
            cout << "true" << endl;

        } else {
            cout << "false" << endl;
        }
        cout << "~"
             << "\n";
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Easy/Strings Rotations of Each Other/strings-rotations-of-each-other.js
//{ Driver Code Starts
// Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let input_line = readLine().split(' ');
        let s1 = input_line[0];
        input_line = readLine().split(' ');
        let s2 = input_line[0];
        let obj = new Solution();
        if (obj.areRotations(s1, s2))
            console.log("true");
        else
            console.log("false");

        console.log("~");
    }
}

// } Driver Code Ends


// User function Template for javascript
/**
 * @param {string} s1
 * @param {string} s2
 * @returns {boolean}
 */

class Solution {
    // Function to check if two strings are rotations of each other or not.
    areRotations(s1, s2) {
        // code here
        
         if(s1.length!=s2.length)
     {
         return false;
     }
     let data = s1+s1;
    //  console.log("data",data);
     return data.includes(s2);
    }
}